src/main/java/fi/aalto/cs/apluscourses/intellij/model/IntelliJModelFactory.java

```diff
@@ -3,43 +3,52 @@
 import com.intellij.ProjectTopics;
 import com.intellij.openapi.project.ModuleListener;
 import com.intellij.openapi.project.Project;
+import com.intellij.openapi.roots.libraries.LibraryTable;
+import com.intellij.openapi.vfs.VirtualFile;
 import com.intellij.openapi.vfs.VirtualFileManager;
 import com.intellij.openapi.vfs.newvfs.BulkFileListener;
 import com.intellij.openapi.vfs.newvfs.events.VFileDeleteEvent;
 import com.intellij.openapi.vfs.newvfs.events.VFileEvent;
 import fi.aalto.cs.apluscourses.model.Course;
+import fi.aalto.cs.apluscourses.model.Library;
 import fi.aalto.cs.apluscourses.model.ModelFactory;
 import fi.aalto.cs.apluscourses.model.Module;
 import java.net.URL;
 import java.util.List;
 import java.util.Map;
+import java.util.Objects;
 import org.jetbrains.annotations.NotNull;
 
 public class IntelliJModelFactory implements ModelFactory {
-
   @NotNull
-  private final Project project;
+  private final APlusProject project;
 
   public IntelliJModelFactory(@NotNull Project project) {
-    this.project = project;
+    this.project = new APlusProject(project);
   }
 
   @Override
   public Course createCourse(@NotNull String name,
                              @NotNull List<Module> modules,
+                             @NotNull List<Library> libraries,
                              @NotNull Map<String, String> requiredPlugins,
                              @NotNull Map<String, URL> resourceUrls) {
-    IntelliJCourse course
-        = new IntelliJCourse(name, modules, requiredPlugins, resourceUrls, project);
+    IntelliJCourse course =
+        new IntelliJCourse(name, modules, libraries, requiredPlugins, resourceUrls, project);
     // Add a module change listener with the created course instance to the project
+    // TODO: These should be handled in those classes; also, renaming issue?
```

Comment:
What should be handled and in which classes? Do you mean the listeners should be added in the classes themselves, e.g. `IntelliJCourse` or...?

Reply:
Yeah, that was my intention. I think it's problematic if the responsibilities to listen all these thing are loaded to this single method. Other ideas?